drivers: serial : uart_nrf5: add power management control

ifillonneau opened the pull request:
NRF52 uart IP seems to consume 500uA when it is activated.
        During low power sessions, we would like to desactivate uart totally.
        Thanks to this commit, power consumption decreases significantly (only few uA...)  as soon as low power control has been requested, 

Signed-off-by: Ismael Fillonneau <ismael.fillonneau@stimio.fr>

mike-scott commented:
@carlescufi @anangl is some of the higher power usage due to using UART mode and not UARTE (EasyDMA)?

carlescufi commented:
@jarz-nordic can you please take a look at this? should @ifillonneau rebase or do you plan to push similar functionality soon?

jarz-nordic commented:
Please rebase. File uart_nrf5.c was renamed to uart_nrfx_uart.c and it was significantly changed.

ifillonneau committed:
drivers: serial : uart_nrfx: add power management control

NRF52 uart IP seems to consume 500uA when it is activated.
During low power sessions, we would like to desactivate uart totally.
Thanks to this patch, power consumption decreases significantly
(only few uA...) as soon as low power control has been requested.

Signed-off-by: Ismael Fillonneau <ismael.fillonneau@stimio.fr>

ifillonneau commented:
I rebased the commit. I saw in latest modifications that you removed "uart_nrf5_dev_data_t" structure,
so I put "device_power_state" variable as global.

carlescufi commented:
@ifillonneau would you mind rebasing instead of merging?  we don't allow merge commits in Pull Requests. More info here: http://docs.zephyrproject.org/contribute/contribute_guidelines.html#contribution-workflow

ifillonneau commented:
online changes was not the good way! new commit has been pushed ...

carlescufi commented:
@jarz-nordic @anangl can you please review this when you have a moment?

carlescufi commented:
@jarz-nordic @anangl Since this has been brought up in the mailing list, can we fast-track the review? Thanks!

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -309,6 +370,8 @@ static void uart_nrfx_isr(void *arg)
 		m_irq_callback(dev);
 	}
+
+	device_busy_clear(dev);
```
This cannot be called here, since `uart_nrfx_fifo_fill` is supposed to be called from the callback. In the current form, right after some byte is passed to the transmitter (by `uart_nrfx_fifo_fill`), the device is marked as not busy, although the transmission of the byte may still be in progress. Instead, the busy status should be cleared when the next IRQ occurs and nothing more is passed to the transmitter (i.e. `uart_nrfx_fifo_fill` is not called again). Thus, the clearing should be done in `uart_nrfx_irq_tx_disable`. Consequently, it would be better to call `device_busy_set` from `uart_nrfx_irq_tx_enable`, not from `uart_nrfx_fifo_fill`.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -199,6 +258,8 @@ static int uart_nrfx_fifo_read(struct device *dev,
 {
 	u8_t num_rx = 0;
 
+	device_busy_set(dev);
```
I see no point in calling it here. For this to work properly, the busy status would need to be cleared before this function returns (now it is cleared at the exit from the interrupt callback, but as I explained in the previous comment this needs to be changed). But is there a need to mark this device as busy only for the time while RX interrupt is being served?

pizi-nordic commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -98,6 +102,59 @@ static int baudrate_set(struct device *dev, u32_t baudrate)
+				void *context)
+{
+	if (ctrl_command == DEVICE_PM_SET_POWER_STATE) {
+		if (*((u32_t *)context) == DEVICE_PM_LOW_POWER_STATE) {
```
In my opinion a switch/case will look much better here.

pizi-nordic commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -98,6 +102,59 @@ static int baudrate_set(struct device *dev, u32_t baudrate)
+			return uart_nrfx_low_power_device(dev);
+		} else if (*((u32_t *)context) == DEVICE_PM_ACTIVE_STATE) {
+			return uart_nrfx_resume_device_from_low_power(dev);
+		}
```
In the proposed implementation the completely disables UART when `DEVICE_PM_LOW_POWER_STATE` is requested, making it no operational. I am not sure, if this power level corresponds with taken action.

pizi-nordic commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -180,6 +237,8 @@ static int uart_nrfx_fifo_fill(struct device *dev,
 {
 	u8_t num_tx = 0;
 
+	device_busy_set(dev);
```
The device_busy_set/clear just sets the flag. Regardless of number of set() call, single clear() clears busy status.
This means that we have to track HW and driver state in the software, as we have to distinguish busy driver (API during execution) and busy hardware (data being sent) in order to clear the busy flag only when both driver and hardware is idle.

MariuszSkamra commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -98,6 +102,59 @@ static int baudrate_set(struct device *dev, u32_t baudrate)
+		return -EBUSY;
+	}
+
+	nrf_uart_disable(NRF_UART0);
```
I think it may cause, the poll hang. There is a while loop in uart_nrfx_poll_out function
 `while (!nrf_uart_event_check(NRF_UART0, NRF_UART_EVENT_TXDRDY)) {}`
that will not exit if TXRDY event bit is not set. And it will not be set if UART has been disabled.
Am I right @anangl ?

anangl replied:
Yes. Although the UART is enabled again at resuming from the low power mode, the `STARTTX` task is not triggered at that point. And this will result in hanging in `_poll_out`.

cvinayak replied:
How is Rx handled in low power state? interrupt-driven and for RTS/CTS flow control lines?

nashif commented:
is this PR still needed?

carlescufi commented:
Yes, this was implemented as part of https://github.com/zephyrproject-rtos/zephyr/pull/9119.

@ifillonneau thanks very much for the contribution.

TSonono commented:
I am encountering sleep currents of ~0.5 mA on nRF52840DK with Zephyr 2.5.0. I am running the system off example which should be disabling the power of the UART peripheral while being in System ON sleep.  At this stage of the application, it is consuming ~0.5 mA for me.

When it goes to System OFF, I get an average of 12 uA.

Could anything have been introduced that messes with this implementation?
